# Importing, Reading and Manipulating Data with ACTUAL LITERAL PANDAS

![I have no idea what I'm doing panda](https://cdn-images-1.medium.com/max/1600/1*oBx032ncOwLmCFX3Epo3Zg.jpeg)

Just kidding - but Pandas is a great library to work with relational data. 

[Check out the documentation!](https://pandas.pydata.org/pandas-docs/stable/) (always a great idea)

Pandas is a library we're going to be using pretty much every day in this course, so we're going to do a ton of practice so you can be on your way to becoming a _PANDAS MASTER_.

![Kung fu panda excited](https://data.whicdn.com/images/201331793/original.gif)

Today, we're going to look at some data from the Austin Animal Shelter. First, let's look at intake data, which is data describing the animals as they enter the shelter.

In [1]:
# Imports! Can't use pandas unless we bring it into our notebook
import pandas as pd

In [2]:
# So... how do we access our data?
# I've provided the path - replace None with appropriate code
PATH = "data/Austin_Animal_Center_Intakes.csv"

intakes = pd.read_csv(PATH)

In [3]:
# Code here to check out the first few rows
intakes.head()

Animal ID     Name                DateTime               MonthYear  \
0   A786884   *Brock  01/03/2019 04:19:00 PM  01/03/2019 04:19:00 PM   
1   A706918    Belle  07/05/2015 12:59:00 PM  07/05/2015 12:59:00 PM   
2   A724273  Runster  04/14/2016 06:43:00 PM  04/14/2016 06:43:00 PM   
3   A665644      NaN  10/21/2013 07:59:00 AM  10/21/2013 07:59:00 AM   
4   A682524      Rio  06/29/2014 10:38:00 AM  06/29/2014 10:38:00 AM   

                        Found Location Intake Type Intake Condition  \
0  2501 Magin Meadow Dr in Austin (TX)       Stray           Normal   
1     9409 Bluegrass Dr in Austin (TX)       Stray           Normal   
2   2818 Palomino Trail in Austin (TX)       Stray           Normal   
3                          Austin (TX)       Stray             Sick   
4        800 Grove Blvd in Austin (TX)       Stray           Normal   

  Animal Type Sex upon Intake Age upon Intake  \
0         Dog   Neutered Male         2 years   
1         Dog   Spayed Female         8 years   
2         Dog     Intact Male       11 months   
3         Cat   Intact Female         4 weeks   
4         Dog   Neutered Male         4 years   

                                   Breed        Color  
0                             Beagle Mix     Tricolor  
1               English Springer Spaniel  White/Liver  
2                            Basenji Mix  Sable/White  
3                 Domestic Shorthair Mix       Calico  
4  Doberman Pinsch/Australian Cattle Dog     Tan/Gray

In [4]:
# Code here to check the shape of the dataframe
intakes.shape

(115089, 12)

In [5]:
# Code here to check the size of the dataframe
intakes.size

1381068

In [6]:
# Code here to check information on the dataframe
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115089 entries, 0 to 115088
Data columns (total 12 columns):
Animal ID           115089 non-null object
Name                78859 non-null object
DateTime            115089 non-null object
MonthYear           115089 non-null object
Found Location      115089 non-null object
Intake Type         115089 non-null object
Intake Condition    115089 non-null object
Animal Type         115089 non-null object
Sex upon Intake     115088 non-null object
Age upon Intake     115089 non-null object
Breed               115089 non-null object
Color               115089 non-null object
dtypes: object(12)
memory usage: 10.5+ MB


Questions to ask yourself:

- How can you describe the shape of the data?
- Are there any empty cells? How can I tell?
- What are the datatypes of the data?

**Note:** You see that some of the ways we dealt with our dataframe required `()` and some did not - why is that?

## Dealing with Null Data

It is a fact of the data science life - you will always be surrounded by 'dirty' data. What does it mean for data to be 'dirty'? What are some of the various ways that data can be 'dirty'?

- 

In [7]:
# Check for null values recognized by pandas as blank
# Run this cell without changes - is this helpful?
intakes.isna()

Animal ID   Name  DateTime  MonthYear  Found Location  Intake Type  \
0           False  False     False      False           False        False   
1           False  False     False      False           False        False   
2           False  False     False      False           False        False   
3           False   True     False      False           False        False   
4           False  False     False      False           False        False   
...           ...    ...       ...        ...             ...          ...   
115084      False   True     False      False           False        False   
115085      False   True     False      False           False        False   
115086      False   True     False      False           False        False   
115087      False  False     False      False           False        False   
115088      False  False     False      False           False        False   

        Intake Condition  Animal Type  Sex upon Intake  Age upon Intake  \
0                  False        False            False            False   
1                  False        False            False            False   
2                  False        False            False            False   
3                  False        False            False            False   
4                  False        False            False            False   
...                  ...          ...              ...              ...   
115084             False        False            False            False   
115085             False        False            False            False   
115086             False        False            False            False   
115087             False        False            False            False   
115088             False        False            False            False   

        Breed  Color  
0       False  False  
1       False  False  
2       False  False  
3       False  False  
4       False  False  
...       ...    ...  
115084  False  False  
115085  False  False  
115086  False  False  
115087  False  False  
115088  False  False  

[115089 rows x 12 columns]

In [8]:
# Code here for a more helpful null check
intakes.isna().sum()

Animal ID               0
Name                36230
DateTime                0
MonthYear               0
Found Location          0
Intake Type             0
Intake Condition        0
Animal Type             0
Sex upon Intake         1
Age upon Intake         0
Breed                   0
Color                   0
dtype: int64

There is no one way to deal with null values. What are some of the strategies we can use to deal with them?

- 

How, in Pandas, can we fill null values recognized by Pandas as null? Let's practice by filling nulls for the Name column with some placeholder value, like 'No name'.

Helpful link: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html

In [9]:
# Code here to fill nulls in the Name column
intakes['Name'].fillna(value='No name')

0          *Brock
1           Belle
2         Runster
3         No name
4             Rio
           ...   
115084    No name
115085    No name
115086    No name
115087     Rianna
115088      *Toad
Name: Name, Length: 115089, dtype: object

Now let's check for nulls again...

In [10]:
# Copy/paste code from above to re-check for nulls - a sanity check!
intakes.isna().sum()

Animal ID               0
Name                36230
DateTime                0
MonthYear               0
Found Location          0
Intake Type             0
Intake Condition        0
Animal Type             0
Sex upon Intake         1
Age upon Intake         0
Breed                   0
Color                   0
dtype: int64

In [11]:
# Actually filling the nulls
intakes['Name'].fillna(value='No name', inplace=True)

Let's try a different strategy for the one lonely null in the 'Sex upon Intake' column - let's just drop that row, since it's only one observation.

Helpful link: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html

In [12]:
# Code here to drop the whole row where Sex upon Intake is null
intakes.dropna(axis=0, how='any', inplace=True)

In [13]:
# Copy/paste code from above to re-check for nulls
intakes.isna().sum()

Animal ID           0
Name                0
DateTime            0
MonthYear           0
Found Location      0
Intake Type         0
Intake Condition    0
Animal Type         0
Sex upon Intake     0
Age upon Intake     0
Breed               0
Color               0
dtype: int64

How do we find sneaky null values that aren't marked by Pandas as null?

In [14]:
# Run this cell without changes
intakes['Age upon Intake'].value_counts()

1 year       20496
2 years      17152
1 month      11091
3 years       6929
2 months      6142
4 years       4189
4 weeks       4126
5 years       3811
3 weeks       3404
4 months      2976
3 months      2967
5 months      2872
6 years       2562
2 weeks       2349
6 months      2235
7 years       2176
8 years       2130
7 months      1756
9 months      1737
10 years      1692
8 months      1395
9 years       1240
1 week         945
10 months      913
12 years       828
1 weeks        788
11 months      724
0 years        702
11 years       693
1 day          561
13 years       524
3 days         500
2 days         455
14 years       354
15 years       307
4 days         302
5 weeks        302
6 days         284
5 days         173
16 years       133
17 years        74
18 years        46
19 years        24
20 years        17
22 years         4
-1 years         4
-3 years         1
25 years         1
21 years         1
24 years         1
Name: Age upon Intake, dtype: int64

Analyze the values you're finding in the 'Age upon Intake' column. What doesn't quite fit here?

**Note:** using `.value_counts()` is just one way to look at the values of a column. In this case, it works because we can see which values are the most common, and it's verbose enough to show even the less common values that might be problematic.

So - how do we want to deal with the data in here that doesn't make sense?

In [15]:
# Code here to fix data??
# Might not want to fix it now - can't find the mean/median until it's on a 
# standard scale and not a string - but discuss now what we might WANT to do

A strategy for dealing with data involves making it so that we can sort by age, and have a standard scale for age.

First, let's see what that would look like if we try it as the column is now:

In [16]:
# Run this cell without changes
intakes['Age upon Intake'].sort_values(ascending=True).unique()

array(['-1 years', '-3 years', '0 years', '1 day', '1 month', '1 week',
       '1 weeks', '1 year', '10 months', '10 years', '11 months',
       '11 years', '12 years', '13 years', '14 years', '15 years',
       '16 years', '17 years', '18 years', '19 years', '2 days',
       '2 months', '2 weeks', '2 years', '20 years', '21 years',
       '22 years', '24 years', '25 years', '3 days', '3 months',
       '3 weeks', '3 years', '4 days', '4 months', '4 weeks', '4 years',
       '5 days', '5 months', '5 weeks', '5 years', '6 days', '6 months',
       '6 years', '7 months', '7 years', '8 months', '8 years',
       '9 months', '9 years'], dtype=object)

Let's unpack what is happening in that line of code - I take the column 'Age upon Intake' by itself (as a series), then sort the values from lowest to highest (`ascending=True`), then grab only unique results so we can see how it ordered the values without looking through all 115,088.

Does that do what we want it to? Let's discuss how this worked - how did it sort?

- 

To make our problem a bit easier, without dealing with the different ways that age is broken out, let's only look at animals where the age is given in years. How can we do that?

In [55]:
# Code here to grab only the animals where age is given in years
year_df = intakes.loc[intakes['Age upon Intake'].str.contains("year") == True]

In [56]:
year_df.head()

Animal ID     Name                DateTime               MonthYear  \
0   A786884   *Brock  01/03/2019 04:19:00 PM  01/03/2019 04:19:00 PM   
1   A706918    Belle  07/05/2015 12:59:00 PM  07/05/2015 12:59:00 PM   
4   A682524      Rio  06/29/2014 10:38:00 AM  06/29/2014 10:38:00 AM   
5   A743852     Odin  02/18/2017 12:46:00 PM  02/18/2017 12:46:00 PM   
6   A635072  Beowulf  04/16/2019 09:53:00 AM  04/16/2019 09:53:00 AM   

                        Found Location      Intake Type Intake Condition  \
0  2501 Magin Meadow Dr in Austin (TX)            Stray           Normal   
1     9409 Bluegrass Dr in Austin (TX)            Stray           Normal   
4        800 Grove Blvd in Austin (TX)            Stray           Normal   
5                          Austin (TX)  Owner Surrender           Normal   
6  415 East Mary Street in Austin (TX)    Public Assist           Normal   

  Animal Type Sex upon Intake Age upon Intake  \
0         Dog   Neutered Male         2 years   
1         Dog   Spayed Female         8 years   
4         Dog   Neutered Male         4 years   
5         Dog   Neutered Male         2 years   
6         Dog   Neutered Male         6 years   

                                   Breed        Color  
0                             Beagle Mix     Tricolor  
1               English Springer Spaniel  White/Liver  
4  Doberman Pinsch/Australian Cattle Dog     Tan/Gray  
5                 Labrador Retriever Mix    Chocolate  
6                         Great Dane Mix        Black

In [57]:
# Check the shape of this subset dataframe
year_df.shape

(66091, 12)

In [58]:
# Sanity check
year_df['Age upon Intake'].unique()

array(['2 years', '8 years', '4 years', '6 years', '14 years', '18 years',
       '1 year', '5 years', '3 years', '15 years', '7 years', '12 years',
       '10 years', '9 years', '0 years', '11 years', '13 years',
       '17 years', '19 years', '16 years', '20 years', '22 years',
       '21 years', '-1 years', '25 years', '24 years', '-3 years'],
      dtype=object)

Can we grab only the number of years from this? Let's make a new column where we can put this data.

In [59]:
# Code here to make a new column, 'Age in Years'
year_df['Age in Years'] = year_df['Age upon Intake'].str.split(" ", expand=True)[0]

# Did you get a 'SettingWithCopyWarning'? No worries - let's discuss

/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [60]:
# Code here to transform that column to an integer
year_df['Age in Years'] = year_df['Age in Years'].astype(int)

/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [61]:
# Code here to check your work
year_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66091 entries, 0 to 115087
Data columns (total 13 columns):
Animal ID           66091 non-null object
Name                66091 non-null object
DateTime            66091 non-null object
MonthYear           66091 non-null object
Found Location      66091 non-null object
Intake Type         66091 non-null object
Intake Condition    66091 non-null object
Animal Type         66091 non-null object
Sex upon Intake     66091 non-null object
Age upon Intake     66091 non-null object
Breed               66091 non-null object
Color               66091 non-null object
Age in Years        66091 non-null int64
dtypes: int64(1), object(12)
memory usage: 7.1+ MB


In [62]:
# Code here to check some statistics on our now-numeric column
year_df['Age in Years'].describe()

count    66091.000000
mean         3.422781
std          3.169181
min         -3.000000
25%          1.000000
50%          2.000000
75%          5.000000
max         25.000000
Name: Age in Years, dtype: float64

In [63]:
# Code here to check the unique values - in order!
year_df['Age in Years'].sort_values().unique()

array([-3, -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
       15, 16, 17, 18, 19, 20, 21, 22, 24, 25])

In [64]:
# Now we can deal with those outliers/nonsense numbers! How?
# First let's check the mean
year_df['Age in Years'].mean()

3.422780711443313

In [65]:
# Now let's check the median
year_df['Age in Years'].median()

2.0

Let's discuss this column - what does it mean that the mean and median are so far off? How will that change if we remove some of the nonsense numbers?

- 

In [68]:
# Code here to deal with those nonsense numbers
year_df["Age in Years"].replace(to_replace=[0, -1, -3], value=2, inplace=True)

/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/generic.py:6786: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [69]:
# Sanity check
year_df["Age in Years"].sort_values().unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 24, 25])

In [70]:
# Code here to re-check your mean/median values

print(year_df['Age in Years'].mean())
print(year_df['Age in Years'].median())

3.4442813696267267
2.0


## Dealing with Duplicates

How can we check for duplicates?

In [78]:
intakes.shape

(115088, 12)

In [27]:
# Code here to check for duplicates
intakes.duplicated().sum()

15

In [74]:
duplicate_intakes = intakes.loc[intakes.duplicated(keep=False) == True]

In [75]:
duplicate_intakes.head(20)

Animal ID     Name                DateTime               MonthYear  \
4985    A696688     Mari  02/10/2015 11:00:00 AM  02/10/2015 11:00:00 AM   
4986    A696688     Mari  02/10/2015 11:00:00 AM  02/10/2015 11:00:00 AM   
6505    A727043   *Larry  05/17/2016 03:46:00 PM  05/17/2016 03:46:00 PM   
6506    A727043   *Larry  05/17/2016 03:46:00 PM  05/17/2016 03:46:00 PM   
8488    A761936  No name  11/12/2017 11:16:00 AM  11/12/2017 11:16:00 AM   
8489    A761936  No name  11/12/2017 11:16:00 AM  11/12/2017 11:16:00 AM   
11276   A561806    Dasia  06/05/2017 11:36:00 AM  06/05/2017 11:36:00 AM   
11277   A561806    Dasia  06/05/2017 11:36:00 AM  06/05/2017 11:36:00 AM   
19295   A773428  *Atreyu  05/31/2018 06:38:00 PM  05/31/2018 06:38:00 PM   
19296   A773428  *Atreyu  05/31/2018 06:38:00 PM  05/31/2018 06:38:00 PM   
37436   A735395      Owl  09/23/2016 12:00:00 PM  09/23/2016 12:00:00 PM   
37437   A735395      Owl  09/23/2016 12:00:00 PM  09/23/2016 12:00:00 PM   
43982   A698864  No name  03/19/2015 01:37:00 PM  03/19/2015 01:37:00 PM   
43983   A698864  No name  03/19/2015 01:37:00 PM  03/19/2015 01:37:00 PM   
44906   A764464   Pepper  09/17/2018 03:09:00 PM  09/17/2018 03:09:00 PM   
44907   A764464   Pepper  09/17/2018 03:09:00 PM  09/17/2018 03:09:00 PM   
47282   A668183    Nacha  06/27/2014 11:34:00 AM  06/27/2014 11:34:00 AM   
47283   A668183    Nacha  06/27/2014 11:34:00 AM  06/27/2014 11:34:00 AM   
48207   A681446  No name  06/16/2014 04:38:00 PM  06/16/2014 04:38:00 PM   
48208   A681446  No name  06/16/2014 04:38:00 PM  06/16/2014 04:38:00 PM   

                          Found Location      Intake Type Intake Condition  \
4985                         Austin (TX)  Owner Surrender           Normal   
4986                         Austin (TX)  Owner Surrender           Normal   
6505    9515 N Lamar Blvd in Austin (TX)            Stray           Normal   
6506    9515 N Lamar Blvd in Austin (TX)            Stray           Normal   
8488   8400 Old Bee Caves in Austin (TX)            Stray           Normal   
8489   8400 Old Bee Caves in Austin (TX)            Stray           Normal   
11276      2002 Nightview in Austin (TX)            Stray           Normal   
11277      2002 Nightview in Austin (TX)            Stray           Normal   
19295    7Th And Gonzalez in Austin (TX)            Stray           Normal   
19296    7Th And Gonzalez in Austin (TX)            Stray           Normal   
37436                        Austin (TX)  Owner Surrender           Normal   
37437                        Austin (TX)  Owner Surrender           Normal   
43982      12129 Glass Rd in Travis (TX)         Wildlife          Injured   
43983      12129 Glass Rd in Travis (TX)         Wildlife          Injured   
44906                     Del Valle (TX)  Owner Surrender           Normal   
44907                     Del Valle (TX)  Owner Surrender           Normal   
47282                        Austin (TX)    Public Assist           Normal   
47283                        Austin (TX)    Public Assist           Normal   
48207      6208 Tupelo Dr in Austin (TX)         Wildlife           Normal   
48208      6208 Tupelo Dr in Austin (TX)         Wildlife           Normal   

      Animal Type Sex upon Intake Age upon Intake  \
4985          Cat   Spayed Female         3 years   
4986          Cat   Spayed Female         3 years   
6505          Cat     Intact Male          1 year   
6506          Cat     Intact Male          1 year   
8488          Dog     Intact Male        9 months   
8489          Dog     Intact Male        9 months   
11276         Dog   Spayed Female         8 years   
11277         Dog   Spayed Female         8 years   
19295         Cat     Intact Male         3 weeks   
19296         Cat     Intact Male         3 weeks   
37436         Cat   Intact Female       10 months   
37437         Cat   Intact Female       10 months   
43982       Other         Unknown          1 year   
43983       Other         Un

In [79]:
# Code here to handle duplicates
intakes.drop_duplicates(inplace=True)

In [81]:
# Sanity check
intakes.shape

(115073, 12)

In [82]:
intakes.duplicated().sum()

0

## Merging Dataframes

We were given two dataframes here - both an Intakes and an Outcomes CSV. Let's merge them!

![Merge diagram from Data Science Made Simple](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png)

[Image from Data Science Made Simple's post on Joining/Merging Pandas Data Frames](http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/)

In [84]:
intakes.shape

(115073, 12)

In [85]:
# Read in our outcomes csv
outcomes = pd.read_csv("data/Austin_Animal_Center_Outcomes.csv")

In [89]:
# Check out our outcomes data
outcomes.head()

Animal ID        Name                DateTime               MonthYear  \
0   A789027      Lennie  02/17/2019 11:44:00 AM  02/17/2019 11:44:00 AM   
1   A720371       Moose  02/13/2016 05:59:00 PM  02/13/2016 05:59:00 PM   
2   A674754         NaN  03/18/2014 11:47:00 AM  03/18/2014 11:47:00 AM   
3   A689724  *Donatello  10/18/2014 06:52:00 PM  10/18/2014 06:52:00 PM   
4   A680969       *Zeus  08/05/2014 04:59:00 PM  08/05/2014 04:59:00 PM   

  Date of Birth Outcome Type Outcome Subtype Animal Type Sex upon Outcome  \
0    02/13/2017     Adoption             NaN         Dog    Neutered Male   
1    10/08/2015     Adoption             NaN         Dog    Neutered Male   
2    03/12/2014     Transfer         Partner         Cat      Intact Male   
3    08/01/2014     Adoption             NaN         Cat    Neutered Male   
4    06/03/2014     Adoption             NaN         Cat    Neutered Male   

  Age upon Outcome                               Breed               Color  
0          2 years             Chihuahua Shorthair Mix               Cream  
1         4 months  Anatol Shepherd/Labrador Retriever                Buff  
2           6 days              Domestic Shorthair Mix        Orange Tabby  
3         2 months              Domestic Shorthair Mix               Black  
4         2 months              Domestic Shorthair Mix  White/Orange Tabby

In [91]:
outcomes.shape

(115213, 12)

What column should we use to merge these DataFrames?

-

In [92]:
intakes['Animal ID'].nunique()

102915

In [93]:
outcomes['Animal ID'].nunique()

103095

In [94]:
# Code here to merge dataframes
combined_df = intakes.merge(outcomes, how='inner', on="Animal ID", 
                            suffixes=("_Intake", "_Outcome"),
                            indicator=True)

In [95]:
# Code here to check out the details of our new dataframe
combined_df.shape

(148246, 24)

In [36]:
combined_df.head()

Animal ID Name_Intake         DateTime_Intake        MonthYear_Intake  \
0   A786884      *Brock  01/03/2019 04:19:00 PM  01/03/2019 04:19:00 PM   
1   A706918       Belle  07/05/2015 12:59:00 PM  07/05/2015 12:59:00 PM   
2   A724273     Runster  04/14/2016 06:43:00 PM  04/14/2016 06:43:00 PM   
3   A665644     No name  10/21/2013 07:59:00 AM  10/21/2013 07:59:00 AM   
4   A682524         Rio  06/29/2014 10:38:00 AM  06/29/2014 10:38:00 AM   

                        Found Location Intake Type Intake Condition  \
0  2501 Magin Meadow Dr in Austin (TX)       Stray           Normal   
1     9409 Bluegrass Dr in Austin (TX)       Stray           Normal   
2   2818 Palomino Trail in Austin (TX)       Stray           Normal   
3                          Austin (TX)       Stray             Sick   
4        800 Grove Blvd in Austin (TX)       Stray           Normal   

  Animal Type_Intake Sex upon Intake Age upon Intake  ...  \
0                Dog   Neutered Male         2 years  ...   
1                Dog   Spayed Female         8 years  ...   
2                Dog     Intact Male       11 months  ...   
3                Cat   Intact Female         4 weeks  ...   
4                Dog   Neutered Male         4 years  ...   

        MonthYear_Outcome Date of Birth     Outcome Type Outcome Subtype  \
0  01/08/2019 03:11:00 PM    01/03/2017         Transfer         Partner   
1  07/05/2015 03:13:00 PM    07/05/2007  Return to Owner             NaN   
2  04/21/2016 05:17:00 PM    04/17/2015  Return to Owner             NaN   
3  10/21/2013 11:39:00 AM    09/21/2013         Transfer         Partner   
4  07/02/2014 02:16:00 PM    06/29/2010  Return to Owner             NaN   

  Animal Type_Outcome Sex upon Outcome Age upon Outcome  \
0                 Dog    Neutered Male          2 years   
1                 Dog    Spayed Female          8 years   
2                 Dog    Neutered Male           1 year   
3                 Cat    Intact Female          4 weeks   
4                 Dog    Neutered Male          4 years   

                           Breed_Outcome Color_Outcome _merge  
0                             Beagle Mix      Tricolor   both  
1               English Springer Spaniel   White/Liver   both  
2                            Basenji Mix   Sable/White   both  
3                 Domestic Shorthair Mix        Calico   both  
4  Doberman Pinsch/Australian Cattle Dog      Tan/Gray   both  

[5 rows x 24 columns]

Is the shape of the merged DataFrame what we expected? Let's discuss.

- 

## Level Up!

1. Find the **age in days** for all animals, not just the ones whose age is provided in years. Be sure to do this on the original dataframe, not just on subsets of the dataframe.

   - (Assume a year is 365 days, and a month is 30 days)

        
2. Ask a few questions of the combined dataframe that you couldn't figure out by just looking at the intakes or outcomes dataframes by themselves.

   - Example: Can you find out how long each animal in the combined dataframe has been in the shelter? 
        
       - Hint: Check out Date Time objects - a new data type that isn't a string or an integer, but which Pandas can recognize as time! https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html

In [99]:
# Code here to answer level up #1

In [100]:
# COde here to answer level up #2